Drivers/Src/Minions.c

```diff
@@ -0,0 +1,408 @@
+    initTxBuf[2] = 0;
+    initRxBuf = 0;
+    ChipSelect();
+    BSP_SPI_Write(initTxBuf, 2);
+    BSP_SPI_Read(&initRxBuf, 1);
+    ChipDeselect();
+    //OR IODIRB to set pin 7 to input and write it back
+    initTxBuf[2] = initRxBuf|0x40;
+    initTxBuf[0]=SPI_OPCODE_W;
+    ChipSelect();
+    BSP_SPI_Write(initTxBuf,3);
+    ChipDeselect();
+
+    switchStatesBitmap = 0x0000;
+
+    OSMutexPost(
+        &CommMutex,
+        OS_OPT_POST_NONE,
+        &err
+    );
+    assertOSError(0,err);
+};
+
+
+/**
+* @brief   Initialize Lights Module
+* @return  void
+*/ 
+static void Lights_Init(void) {
+    OS_ERR err;
+    CPU_TS timestamp;
+    BSP_GPIO_Init(LIGHTS_PORT, 0x3C0, 1); // Pins 6,7,8,9 from Port C are out (0b1111000000)
+    
+    // Initialize txBuf and rxBuf
+    uint8_t txReadBuf[2] = {SPI_OPCODE_R, SPI_IODIRB}; //0x01 is IODIRB in Bank 0 Mode
+    uint8_t rxBuf = 0;
+
+    OSMutexPend(
+        &CommMutex,
+        0,
+        OS_OPT_PEND_BLOCKING,
+        &timestamp,
+        &err);
+    assertOSError(OS_BLINK_LIGHTS_LOC,err);
+    
+    // Reads direction register and stores in rxBuf
+    ChipSelect();
+    BSP_SPI_Write(txReadBuf, 2);
+    BSP_SPI_Read(&rxBuf,1);
+    ChipDeselect();
+
+    uint8_t txWriteBuf[3] = {
+        SPI_OPCODE_W, //write to IODIRB
+        SPI_IODIRB, 
+        rxBuf&0x40 //clear IODIRB to set as outputs except for hazard lights switch pin (bit 6, 0x40)
+    };
+    
+    // Set direction register
+    ChipSelect();
+    BSP_SPI_Write(txWriteBuf, 3);
+    ChipDeselect();
+
+    // Initialize toggle bitmap and states bitmap
+    lightToggleBitmap = 0x0000;
+    lightStatesBitmap = 0x0000;
+    
+    // Unlock mutex
+    OSMutexPost(
+        &CommMutex,
+        OS_OPT_POST_NONE,
+        &err);
+    assertOSError(OS_BLINK_LIGHTS_LOC, err);
+}
+
+/**
+ * @brief   Initializes all Lights and Switches. **Can be called only after Fault Semaphore is initialized.**
+ * @param   None
+ * @return  None
+ */ 
+void Minions_Init(void){
+    OS_ERR err;
+    OSMutexCreate(&CommMutex, "Communications Mutex", &err);
+    assertOSError(0,err);
+    BSP_SPI_Init();
+    Lights_Init();
+    Switches_Init();
+}
+
+/**
+* @brief   Read the state of a specific light from the light bitmap
+* @param   light Which Light to read
+* @return  returns State enum which indicates ON/OFF
+*/ 
+State Lights_Read(light_t light) {
+    //Return from a stored state bitmap instead of actually querying hardware. 
+    return (State) ((lightStatesBitmap>>light)&0x0001);
+}
+
+/**
+* @brief   Read the lights bitmap
+* @return  returns uint16_t with lights bitmap
+*/ 
+uint16_t Lights_Bitmap_Read() {
+    //Return from a stored state array instead of actually querying hardware.
+    return lightStatesBitmap;
+}
+
+/**
+ * @brief   Reads from static variable bitmap holding values of switches
+ * @param   sw
+ * @return  ON/OFF State
+ */ 
+State Switches_Read(switches_t sw){
+    return (State) ((switchStatesBitmap >> sw) & 0x0001);
+}
```

Comment:
@Cam0Cow Do you think we should make Switches_Read pend on the mutex and have UpdateSwitches lock the mutex for bitmap writes so that reads block if a write is going on? I don't know how critical it is considering the worst case is you get a slightly out of date switch reading by 100ms or so based on ReadSwitches

Reply:
100ms being the worst case which is during Array precharge, once array and motor are on then at worst its like a 10ms out of date.

Reply:
Since we poll the switch states super often, I don't think it's going to be an issue. The driver won't notice an action being done 10 ms later every once in a while.